File: code/backend-php/src/Core/Service/Auth/SessionService.php
<?php

namespace Games\Paths\Core\Service\Auth;

use Games\Paths\Core\Domain\Auth\RefreshedSession;
use Games\Paths\Core\Domain\Auth\TokenInfo;
use Games\Paths\Core\Port\Auth\JwtPort;
use Games\Paths\Core\Port\Auth\SessionPort;
use Games\Paths\Core\Port\Auth\TokenPersistencePort;

/**
 * Domain service for session and token management.
 * 
 * Implements Token Rotation:
 * When a refresh token is used, ALL existing sessions for that user are revoked.
 * This is a security measure to detect and mitigate token theft (Step 13).
 */
class SessionService implements SessionPort
{
    private JwtPort $jwtPort;
    private TokenPersistencePort $tokenPersistence;
    private int $maxTokensPerUser;

    public function __construct(
        JwtPort $jwtPort,
        TokenPersistencePort $tokenPersistence,
        int $maxTokensPerUser = 5
    ) {
        $this->jwtPort = $jwtPort;
        $this->tokenPersistence = $tokenPersistence;
        $this->maxTokensPerUser = $maxTokensPerUser;
    }

    public function refreshSession(string $refreshToken): RefreshedSession
    {
        // 1. Validate refresh token structure and signature
        if (!$this->jwtPort->validateToken($refreshToken)) {
            throw new \RuntimeException("INVALID_REFRESH_TOKEN", 401);
        }

        // 2. Parse claims
        $tokenInfo = $this->jwtPort->parseToken($refreshToken);
        if (!$tokenInfo->isRefreshToken()) {
            throw new \RuntimeException("NOT_A_REFRESH_TOKEN", 401);
        }

        // 3. Verify in DB (not revoked)
        if (!$this->tokenPersistence->isRefreshTokenValid($refreshToken)) {
            throw new \RuntimeException("REVOKED_REFRESH_TOKEN", 401);
        }

        // 4. Implement Token Rotation: Revoke ALL sessions for this user
        $this->tokenPersistence->revokeAllByUserUuid($tokenInfo->getUserUuid());

        // 5. Generate new pair
        // We need the username and role for the access token. 
        // In a real scenario, we might want to reload the user from DB,
        // but for Step 13, we trust the claims in the refresh token if it's signed.
        $newAccessToken = $this->jwtPort->generateAccessToken(
            $tokenInfo->getUserUuid(),
            $tokenInfo->getUsername(),
            $tokenInfo->getRole()
        );
        $newRefreshToken = $this->jwtPort->generateRefreshToken($tokenInfo->getUserUuid());
        
        // Parse new refresh token info for storage
        $newRefreshInfo = $this->jwtPort->parseToken($newRefreshToken);

        // 6. Persist new refresh token and enforce limits
        $userId = $this->tokenPersistence->findUserIdByUuid($tokenInfo->getUserUuid());
        if ($userId === null) {
            throw new \RuntimeException("USER_NOT_FOUND", 404);
        }

        $this->tokenPersistence->saveRefreshToken($tokenInfo->getUserUuid(), $newRefreshToken, $newRefreshInfo);
        $this->tokenPersistence->enforceTokenLimit($userId, $this->maxTokensPerUser);

        // 7. Return refreshed session
        return new RefreshedSession(
            $tokenInfo->getUserUuid(),
            $tokenInfo->getUsername(),
            $tokenInfo->getRole(),
            $newAccessToken,
            $newRefreshToken,
            time() + ($this->jwtPort->getAccessTokenExpirationMs() / 1000),
            time() + ($this->jwtPort->getRefreshTokenExpirationMs() / 1000)
        );
    }

    public function logout(string $refreshToken): void
    {
        // We only revoke the specific refresh token
        $this->tokenPersistence->revokeToken($refreshToken);
    }

    public function logoutAll(string $userUuid): void
    {
        // Revoke all refresh tokens for this user
        $this->tokenPersistence->revokeAllByUserUuid($userUuid);
    }

    public function validateAccessToken(string $accessToken): TokenInfo
    {
        if (!$this->jwtPort->validateToken($accessToken)) {
            throw new \RuntimeException("INVALID_ACCESS_TOKEN", 401);
        }

        $tokenInfo = $this->jwtPort->parseToken($accessToken);
        if (!$tokenInfo->isAccessToken()) {
            throw new \RuntimeException("NOT_AN_ACCESS_TOKEN", 401);
        }

        return $tokenInfo;
    }
}

File: code/backend/python/app/core/services/auth/session_service.py
from app.core.ports.auth.session_port import SessionPort
from app.core.ports.auth.jwt_port import JwtPort
from app.core.ports.auth.token_persistence_port import TokenPersistencePort
from app.core.models.auth.refreshed_session import RefreshedSession
from app.core.models.auth.token_info import TokenInfo
import time

class SessionService(SessionPort):
    def __init__(self, jwt_port: JwtPort, token_persistence: TokenPersistencePort, max_tokens_per_user: int = 5):
        self.jwt_port = jwt_port
        self.token_persistence = token_persistence
        self.max_tokens_per_user = max_tokens_per_user

    def refresh_session(self, refresh_token: str) -> RefreshedSession:
        # 1. Validate refresh token
        if not self.jwt_port.validate_token(refresh_token):
            raise ValueError("INVALID_REFRESH_TOKEN")

        # 2. Parse claims
        token_info = self.jwt_port.parse_token(refresh_token)
        if not token_info.is_refresh_token():
            raise ValueError("NOT_A_REFRESH_TOKEN")

        # 3. Verify in DB
        if not self.token_persistence.is_refresh_token_valid(refresh_token):
            raise ValueError("REVOKED_REFRESH_TOKEN")

        # 4. Token Rotation: Revoke ALL
        self.token_persistence.revoke_all_by_user_uuid(token_info.user_uuid)

        # 5. Generate new pair
        new_access_token = self.jwt_port.generate_access_token(
            token_info.user_uuid,
            token_info.username,
            token_info.role
        )
        new_refresh_token = self.jwt_port.generate_refresh_token(token_info.user_uuid)
        new_refresh_info = self.jwt_port.parse_token(new_refresh_token)

        # 6. Persist and enforce limit
        user_id = self.token_persistence.find_user_id_by_uuid(token_info.user_uuid)
        if user_id is None:
            raise ValueError("USER_NOT_FOUND")

        self.token_persistence.save_refresh_token(token_info.user_uuid, new_refresh_token, new_refresh_info)
        self.token_persistence.enforce_token_limit(user_id, self.max_tokens_per_user)

        return RefreshedSession(
            userUuid=token_info.user_uuid,
            username=token_info.username,
            role=token_info.role,
            accessToken=new_access_token,
            refreshToken=new_refresh_token,
            accessTokenExpiresAt=int(time.time() + (self.jwt_port.get_access_token_expiration_ms() / 1000)),
            refreshTokenExpiresAt=int(time.time() + (self.jwt_port.get_refresh_token_expiration_ms() / 1000))
        )

    def logout(self, refresh_token: str) -> None:
        self.token_persistence.revoke_token(refresh_token)

    def logout_all(self, user_uuid: str) -> None:
        self.token_persistence.revoke_all_by_user_uuid(user_uuid)

    def validate_access_token(self, access_token: str) -> TokenInfo:
        if not self.jwt_port.validate_token(access_token):
            raise ValueError("INVALID_ACCESS_TOKEN")

        token_info = self.jwt_port.parse_token(access_token)
        if not token_info.is_access_token():
            raise ValueError("NOT_AN_ACCESS_TOKEN")

        return token_info
